Controller/RedirectToServiceController.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+/*
+ * This file is part of the HWIOAuthBundle package.
+ *
+ * (c) Hardware.Info <opensource@hardware.info>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace HWI\Bundle\OAuthBundle\Controller;
+
+use HWI\Bundle\OAuthBundle\Security\OAuthUtils;
+use RuntimeException;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+/**
+ * @author Alexander <iam.asm89@gmail.com>
+ */
+final class RedirectToServiceController
+{
+    /**
+     * @var OAuthUtils
+     */
+    private $oauthUtils;
+
+    /**
+     * @var array
+     */
+    private $firewallNames;
+
+    /**
+     * @var string
+     */
+    private $targetPathParameter;
+
+    /**
+     * @var bool
+     */
+    private $failedUseReferer;
+
+    /**
+     * @var bool
+     */
+    private $useReferer;
+
+    /**
+     * @param OAuthUtils $oauthUtils
+     * @param array      $firewallNames
+     * @param string     $targetPathParameter
+     * @param bool       $failedUseReferer
+     * @param bool       $useReferer
+     */
+    public function __construct(OAuthUtils $oauthUtils, array $firewallNames, ?string $targetPathParameter, bool $failedUseReferer, bool $useReferer)
+    {
+        $this->oauthUtils = $oauthUtils;
+        $this->firewallNames = $firewallNames;
+        $this->targetPathParameter = $targetPathParameter;
+        $this->failedUseReferer = $failedUseReferer;
+        $this->useReferer = $useReferer;
+    }
+
+    /**
+     * @param Request $request
+     * @param string  $service
+     *
+     * @throws NotFoundHttpException
+     *
+     * @return RedirectResponse
+     */
+    public function redirectToServiceAction(Request $request, $service)
+    {
+        try {
+            $authorizationUrl = $this->oauthUtils->getAuthorizationUrl($request, $service);
+        } catch (RuntimeException $e) {
+            throw new NotFoundHttpException($e->getMessage(), $e);
+        }
+
+        $this->storeReturnPath($request, $authorizationUrl);
+
+        return new RedirectResponse($authorizationUrl);
+    }
+
+    private function storeReturnPath(Request $request, string $authorizationUrl): void
+    {
+        $session = $request->getSession();
+
+        if (null === $session) {
+            return;
+        }
+
+        foreach ($this->firewallNames as $providerKey) {
+            $sessionKey = '_security.'.$providerKey.'.target_path';
+            $sessionKeyFailure = '_security.'.$providerKey.'.failed_target_path';
+
+            $param = $this->targetPathParameter;
```

Comment:
This should be out off loop as it's not being changed between checks.

Reply:
Done